### PR TITLE
Adding tests and logic for DMRG and SHCI state import

### PR DESCRIPTION
**Pull request by Chiffafox**
**Context:**

DMRG and SHCI are two state-of-the-art computational chemistry techniques that we use to generate initial states for quantum algorithms for quantum chemistry.

**Description of the Change:**

This change adds logic for the `import_state` function to support automatic identification of input corresponding to DMRG and SHCI solutions, as well as enhances the tests for these methods.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qchem/convert.py [hunk 1]
```diff
@@ -565,11 +565,13 @@ def _ucisd_state(cisd_solver, tol=1e-15):
 def import_state(solver, tol=1e-15):
     r"""Convert an external wavefunction to a state vector.
 
-    Currently, the only accepted source of wavefunctions is the PySCF library.
-    The restricted and unrestricted CISD/CCSD methods are supported.
+    Currently, the only accepted sources of wavefunctions are a) the PySCF library (the
+    restricted and unrestricted CISD/CCSD methods are supported), b) the library Dice
+    implementing the SHCI method, c) the library Block2 implementing the DMRG method.
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:570]: Great point, will do that!

File: pennylane/qchem/convert.py [hunk 2]
```diff
@@ -602,12 +604,28 @@ def import_state(solver, tol=1e-15):
         wf_dict = _rccsd_state(solver, tol=tol)
     elif "UCCSD" in method:
         wf_dict = _uccsd_state(solver, tol=tol)
+    elif "tuple" in method:
+        if isinstance(solver[0][0], np.str_):
```
Comment (soranjh) [inline pennylane/qchem/convert.py:608]: The determinants must be a **list** (not an **array**) of integers for dmrg and a list of strings for shci.

File: tests/qchem/of_tests/test_convert.py [hunk 3]
```diff
@@ -942,6 +942,68 @@ def test_import_state(molecule, basis, symm, method, wf_ref):
     assert np.allclose(wf_comp, wf_ref) or np.allclose(wf_comp, -wf_ref)
 
 
+@pytest.mark.parametrize(
+    ("detscoeffs", "wf_ref"),
+    [
+        (
+            # dmrg
+            (np.array([[0, 3], [3, 0]]).numpy(), np.array([-0.10660077, 0.9943019]).numpy()),
```
Comment (soranjh) [inline tests/qchem/of_tests/test_convert.py:950]: It should be a list not an array:` [[0, 3], [3, 0]]`

File: tests/qchem/of_tests/test_convert.py [hunk 4]
```diff
@@ -942,6 +942,68 @@ def test_import_state(molecule, basis, symm, method, wf_ref):
...
+                    0.9943019,
+                    0.0,
+                    0.0,
+                    0.0,
+                ]
+            ).numpy(),
+        ),
+        (
+            # shci
+            (np.array(["02", "20"]).numpy(), np.array([-0.1066467, 0.99429698]).numpy()),
```
Comment (soranjh) [inline tests/qchem/of_tests/test_convert.py:974]: It should be a list not an array: ["02", "20"]

File: pennylane/qchem/convert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/convert.py:570]: @Chiffafox You can find some hints [here](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html#unordered-lists).

File: pennylane/qchem/convert.py [hunk 5]
```diff
@@ -602,12 +604,28 @@ def import_state(solver, tol=1e-15):
...
+                " tuple(array[str], array[float]) for SHCI calculations with Dice library and "
+                "tuple(array[int], array[float]) for DMRG calculations with the Block2 library."
+            )
     else:
         raise ValueError(
             "The supported objects are RCISD, UCISD, RCCSD, and UCCSD for restricted and"
-            " unrestricted configuration interaction and coupled cluster calculations."
+            " unrestricted configuration interaction and coupled cluster calculations, and"
+            " tuple(array[str], array[float]) for SHCI calculations with Dice library and "
+            "tuple(array[int], array[float]) for DMRG calculations with the Block2 library."
```
Comment (soranjh) [inline pennylane/qchem/convert.py:623]: Please modify these after using list for determinants.

File: pennylane/qchem/convert.py [hunk 6]
```diff
@@ -1144,10 +1162,10 @@ def _shci_state(wavefunction, tol=1e-15):
     SHCI.outputfile.
 
     Args:
-        wavefunction tuple(array[int], array[str]): determinants and coefficients in physicist notation
+        wavefunction tuple(array[str], array[float]): determinants and coefficients in chemist notation
```
Comment (soranjh) [inline pennylane/qchem/convert.py:1165]: Please modify these after using list for determinants.

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @Chiffafox, left some comments.

File: pennylane/qchem/convert.py [hunk 7]
```diff
@@ -602,12 +604,28 @@ def import_state(solver, tol=1e-15):
...
+        if isinstance(solver[0][0], np.str_):
+            wf_dict = _shci_state(solver, tol=tol)
+        elif isinstance(solver[0][0][0], type(np.array([0]).numpy()[0])):
+            wf_dict = _dmrg_state(solver, tol=tol)
+        else:
+            raise ValueError(
+                "for tuple input, the supported objects are"
+                " tuple(array[str], array[float]) for SHCI calculations with Dice library and "
+                "tuple(array[int], array[float]) for DMRG calculations with the Block2 library."
+            )
```
Comment (soranjh) [inline pennylane/qchem/convert.py:617]: Is this tested?

Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Added a test for this.

File: pennylane/qchem/convert.py (hunk 6 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Done.

File: pennylane/qchem/convert.py (hunk 5 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Done.

File: tests/qchem/of_tests/test_convert.py (hunk 4 above)
Comment (Chiffafox) [reply tests/qchem/of_tests/test_convert.py]: Done.

File: tests/qchem/of_tests/test_convert.py (hunk 3 above)
Comment (Chiffafox) [reply tests/qchem/of_tests/test_convert.py]: Done.

File: pennylane/qchem/convert.py (hunk 2 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Copy that, changed that throughout.

File: pennylane/qchem/convert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: It is now a list.

**Review by soranjh — APPROVED** [review]
Looks good, thanks @Chiffafox!

File: tests/qchem/of_tests/test_convert.py [hunk 8]
```diff
@@ -950,6 +1012,11 @@ def test_import_state_error():
     with pytest.raises(ValueError, match="The supported objects"):
         _ = qchem.convert.import_state(myci)
 
+    mytuple = (np.array([[3, 0], [0, 3]]), np.array([0]))
+
+    with pytest.raises(ValueError, match="For tuple input"):
+        _ = qchem.convert.import_state(mytuple)
```
Comment (DSGuala) [inline tests/qchem/of_tests/test_convert.py:1018]: Do we need `_ = `? Can we call only `qchem.convert.import_state(mytuple)`?

**Review by DSGuala — APPROVED** [review]
Looks good. I think it's ready to merge but I left some comments/questions.

File: tests/qchem/of_tests/test_convert.py (hunk 8 above)
Comment (Chiffafox) [reply tests/qchem/of_tests/test_convert.py]: Thank you for flagging this -- removed this!

File: tests/qchem/of_tests/test_convert.py [hunk 9]
```diff
@@ -942,6 +942,68 @@ def test_import_state(molecule, basis, symm, method, wf_ref):
...
+def test_import_state_nonpyscf(detscoeffs, wf_ref):
+    r"""Test that import_state returns the correct state vector."""
+
+    wf_comp = qchem.convert.import_state(detscoeffs)
+
+    # overall sign could be different in each PySCF run
+    assert np.allclose(wf_comp, wf_ref) or np.allclose(wf_comp, -wf_ref)
+
+
 def test_import_state_error():
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply tests/qchem/of_tests/test_convert.py:1007]: I think with the tuple length check I added above we are okay.

File: pennylane/qchem/convert.py [hunk 10]
```diff
@@ -602,12 +607,28 @@ def import_state(solver, tol=1e-15):
         wf_dict = _rccsd_state(solver, tol=tol)
     elif "UCCSD" in method:
         wf_dict = _uccsd_state(solver, tol=tol)
+    elif "tuple" in method:
+        if isinstance(solver[0][0], str):
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Added length check on the tuple.

File: pennylane/qchem/convert.py [hunk 11]
```diff
@@ -565,11 +565,16 @@ def _ucisd_state(cisd_solver, tol=1e-15):
...
+
+        * the PySCF library (the restricted and unrestricted CISD/CCSD
+          methods are supported),
+        * the library Dice implementing the SHCI method,
+        * the library Block2 implementing the DMRG method.
 
     Args:
-        solver: external wavefunction object (e.g. PySCF Solver object) that will be converted
+        solver: external wavefunction object (e.g. PySCF Solver object) or tuple of
+        determinants and coefficients that will be converted
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Added clarification to this point.

File: pennylane/qchem/convert.py [hunk 12]
```diff
@@ -565,18 +565,26 @@ def _ucisd_state(cisd_solver, tol=1e-15):
...
     r"""Convert an external wavefunction to a state vector.
 
-    Currently, the only accepted source of wavefunctions is the PySCF library.
-    The restricted and unrestricted CISD/CCSD methods are supported.
+    Currently, the only accepted sources of wavefunctions are
+
+        * the PySCF library (the restricted and unrestricted CISD/CCSD
+          methods are supported),
+        * the library Dice implementing the SHCI method,
+        * the library Block2 implementing the DMRG method.
```
Comment (soranjh) [inline pennylane/qchem/convert.py:573]: Better to remove ',' and '.' at the end.

File: pennylane/qchem/convert.py [hunk 13]
```diff
@@ -565,18 +565,26 @@ def _ucisd_state(cisd_solver, tol=1e-15):
...
 
     Args:
-        solver: external wavefunction object (e.g. PySCF Solver object) that will be converted
+        solver: external wavefunction object.
+
+            * For PySCF, this is the R/UCISD or R/UCCSD Solver object,
+            * For SHCI via Dice, this is the tuple(list[str], array[float]) of Slater determinants and their coefficients;
+            * For DMRG via Block2, this is the tuple(list[int], array[float]) of Slater determinants and their coefficients
```
Comment (soranjh) [inline pennylane/qchem/convert.py:580]: Not sure if it looks good in the built docs. Please consider merging this information with the bullet points above.

File: pennylane/qchem/convert.py [hunk 14]
```diff
@@ -565,18 +565,26 @@ def _ucisd_state(cisd_solver, tol=1e-15):
...
         tol (float): the tolerance for discarding Slater determinants based on their coefficients
 
     Raises:
         ValueError: if external object type is not supported
 
     Returns:
-        array: normalized state vector of length ``2**len(number_of_spinorbitals)``
+        array: normalized state vector of length :math:`2^M`, where :math:`M` is the number_of_spinorbitals.
```
Comment (soranjh) [inline pennylane/qchem/convert.py:587]: Please remove `'_'` in number_of_spinorbitals.

File: pennylane/qchem/convert.py (hunk 12 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py:573]: Done.

File: pennylane/qchem/convert.py (hunk 13 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py:580]: Fixed.

File: pennylane/qchem/convert.py (hunk 14 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py:587]: Done.

File: pennylane/qchem/convert.py [hunk 15]
```diff
@@ -17,7 +17,7 @@
 import warnings
 from itertools import product
 
-# pylint: disable=import-outside-toplevel
+# pylint: disable= too-many-branches, too-many-arguments, too-many-locals, too-many-nested-blocks, import-outside-toplevel
```
Comment (trbromley) [inline pennylane/qchem/convert.py:20]: Not a priority for now, but this might be a sign that the function could be modularized a bit.

Comment (soranjh) [reply pennylane/qchem/convert.py:20]: The `import_state` function had `too-many-branches`. The rest were not needed. A fix is [here](https://github.com/PennyLaneAI/pennylane/pull/4634/files) to remove them all.

**Pull request merged**